src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,397 @@
+        private isWebExtension: boolean,
+        private readonly serverUri: IJupyterServerUriEntry
+    ) {
+        // Register with remote-serverId as our ID
+        this.id = `${this.kind}-${serverUri.serverId}`;
+
+        // Create a reasonable display name for this kernel finder
+        this.displayName = localize.DataScience.universalRemoteKernelFinderDisplayName().format(
+            serverUri.displayName || serverUri.uri
+        );
+
+        this._initializedPromise = new Promise<void>((resolve) => {
+            this._initializeResolve = resolve;
+        });
+
+        // When we register, add a disposable to clean ourselves up from the main kernel finder list
+        // Unlike the Local kernel finder universal remote kernel finders will be added on the fly
+        this.disposables.push(kernelFinder.registerKernelFinder(this));
+    }
+
+    dispose(): void | undefined {
+        disposeAllDisposables(this.disposables);
+    }
+
+    async activate(): Promise<void> {
+        // warm up the cache
+        this.loadCache().then(noop, noop);
+
+        // If we create a new kernel, we need to refresh if the kernel is remote (because
+        // we have live sessions possible)
+        // Note, this is a perf optimization for right now. We should not need
+        // to check for remote if the future when we support live sessions on local
+        this.kernelProvider.onDidStartKernel(
+            (k) => {
+                if (isRemoteConnection(k.kernelConnectionMetadata)) {
+                    // update remote kernels
+                    this.updateCache().then(noop, noop);
+                }
+            },
+            this,
+            this.disposables
+        );
+
+        // For kernel dispose we need to wait a bit, otherwise the list comes back the
+        // same
+        this.kernelProvider.onDidDisposeKernel(
+            (k) => {
+                if (k && isRemoteConnection(k.kernelConnectionMetadata)) {
+                    const timer = setTimeout(() => {
+                        this.updateCache().then(noop, noop);
+                    }, REMOTE_KERNEL_REFRESH_INTERVAL);
+
+                    return timer;
+                }
+            },
+            this,
+            this.disposables
+        );
+
+        this.extensions.onDidChange(
+            () => {
+                // If we just installed the Python extension and we fetched the controllers, then fetch it again.
+                if (!this.wasPythonInstalledWhenFetchingKernels && this.extensionChecker.isPythonExtensionInstalled) {
+                    this.updateCache().then(noop, noop);
+                }
+            },
+            this,
+            this.disposables
+        );
+        this.wasPythonInstalledWhenFetchingKernels = this.extensionChecker.isPythonExtensionInstalled;
+    }
+
+    public async loadCache() {
+        traceInfoIfCI(`Remote Kernel Finder load cache Server: ${this.id}`);
```

Comment:
this is an internal  implementation, why do we need to make this public
couldn't find it being used

Reply:
Ahh, looks like just a copy from the existing one. That that was used in unit tests. Cleaned up.